[WebGPU] After an invalid drawIndexed call, some valid draws can be discarded

mwyrzykowski opened the pull request:
#### 9600be358d0babb699b59988f17e59e10025ff4c
<pre>
[WebGPU] After an invalid drawIndexed call, some valid draws can be discarded
<a href="https://bugs.webkit.org/show_bug.cgi?id=275111">https://bugs.webkit.org/show_bug.cgi?id=275111</a>
&lt;radar://129236340&gt;

Reviewed by Dan Glastonbury.

Range check is wrong because a valid call within a previously invalid call
will be treated as a no-op, and alternatively the other way around.

* LayoutTests/fast/webgpu/regression/repro_275111-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_275111.html: Added.
* LayoutTests/fast/webgpu/regression/repro_275111b-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_275111b.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::indirectBufferRequiresRecomputation const):

Canonical link: <a href="https://commits.webkit.org/279750@main">https://commits.webkit.org/279750@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cac0855c39da5ead06bac58d156e44941d111f41

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6952 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57671 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5123 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5071 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44050 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3432 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56485 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47100 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25186 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4424 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3266 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59262 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51475 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47192 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11860 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30560 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
Does this need to be rebased onto your other changes to pick up minInstanceCount?

:shipit:

mwyrzykowski committed:
[WebGPU] After an invalid drawIndexed call, some valid draws can be discarded
https://bugs.webkit.org/show_bug.cgi?id=275111
<radar://129236340>

Reviewed by Dan Glastonbury.

Range check is wrong because a valid call within a previously invalid call
will be treated as a no-op, and alternatively the other way around.

* LayoutTests/fast/webgpu/regression/repro_275111-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_275111.html: Added.
* LayoutTests/fast/webgpu/regression/repro_275111b-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_275111b.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::indirectBufferRequiresRecomputation const):

Canonical link: https://commits.webkit.org/279750@main
